core/src/main/java/org/elasticsearch/common/settings/Setting.java

```diff
@@ -61,32 +65,99 @@
+ * public static final Setting<Color> MY_BOOLEAN =
+ *     new Setting<>("my.color.setting", Color.RED.toString(), Color::valueOf, SettingsProperty.ClusterScope);
  * }
  * </pre>
  */
 public class Setting<T> extends ToXContentToBytes {
+
+    public enum SettingsProperty {
+        /**
+         * should be filtered in some api (mask password/credentials)
+         */
+        Filtered,
+
+        /**
+         * iff this setting can be dynamically updateable
+         */
+        Dynamic,
+
+        /**
+         * mark this setting as deprecated
+         */
+        Deprecated,
+
+        /**
+         * Cluster scope.
+         * @See IndexScope
+         * @See NodeScope
+         */
+        ClusterScope,
+
+        /**
+         * Node scope.
+         * @See ClusterScope
+         * @See IndexScope
+         */
+        NodeScope,
+
+        /**
+         * Index scope.
+         * @See ClusterScope
+         * @See NodeScope
+         */
+        IndexScope;
+    }
+
+    private static final ESLogger logger = Loggers.getLogger(Setting.class);
+    private static final DeprecationLogger deprecationLogger = new DeprecationLogger(logger);
+
     private final String key;
     protected final Function<Settings, String> defaultValue;
     private final Function<String, T> parser;
-    private final boolean dynamic;
-    private final Scope scope;
+    private final EnumSet<SettingsProperty> properties;
 
     /**
      * Creates a new Setting instance
      * @param key the settings key for this setting.
      * @param defaultValue a default value function that returns the default values string representation.
      * @param parser a parser that parses the string rep into a complex datatype.
-     * @param dynamic true iff this setting can be dynamically updateable
-     * @param scope the scope of this setting
+     * @param properties properties for this setting like scope, filtering...
      */
-    public Setting(String key, Function<Settings, String> defaultValue, Function<String, T> parser, boolean dynamic, Scope scope) {
+    public Setting(String key, Function<Settings, String> defaultValue, Function<String, T> parser, SettingsProperty... properties) {
         assert parser.apply(defaultValue.apply(Settings.EMPTY)) != null || this.isGroupSetting(): "parser returned null";
         this.key = key;
         this.defaultValue = defaultValue;
         this.parser = parser;
-        this.dynamic = dynamic;
-        this.scope = scope;
+        if (properties.length == 0) {
+            this.properties = EnumSet.of(SettingsProperty.NodeScope);
+        } else {
+            this.properties = EnumSet.copyOf(Arrays.asList(properties));
+        }
+        // We validate scope settings. They are mutually exclusive
+        int numScopes = 0;
+        for (SettingsProperty property : properties) {
+            if (property == SettingsProperty.ClusterScope ||
```

Comment:
I think we can allow that, we might have settings that have IndexScope but can also be set on a Node level but others not?